Clarify hardware profile and instance configuration related docs for ECH (#2039)
elastic/docs-content: deploy-manage/deploy/elastic-cloud/change-hardware.md

  - id: cloud-hosted
---

> # Customize instance configuration [ec-change-instance-configuration]

Reviewer: Changing the anchor name (the part under `[` and `]` could provoke existing links to this page (using the old anchor name) to fail.

Changing the heading name is not a problem, of course.

I would only change the anchor names if I was sure it's an anchor that hasn't been used or referenced in the past.

> # Customize instance configuration [ec-change-instance-configuration]

Reviewer: Please add an introduction to the document explaining its scope / purpose, and also explaining when the document can be useful (what use case it addresses).

I would start with something like:

> This document explains how to modify the instance configurations used by specific components of your deployment without changing the overall hardware profile assigned to the deployment. This advanced configuration scenario is useful when ...... (to be determined)

We need to explain `when` such a procedure would be needed by our users.

## Consideration [ec-consideration-on-changing-ic]

> The virtual hardware on which {{stack}} deployments run is defined by instance configurations. Review [Hardware profile](./ec-change-hardware-profile.md#ec-hardware-profile) and [Instance configurations](cloud://reference/cloud-hosted/hardware.md#ec-getting-started-configurations) for more information about hardware profile and instance configuration respectively.

Reviewer: I'm suggesting to change this sentence to active voice.

```suggestion
{{stack}} deployments run on virtual hardware defined by instance configurations. For more details, refer to [Hardware profiles](./ec-change-hardware-profile.md#ec-hardware-profile) and [Instance configurations](cloud://reference/cloud-hosted/hardware.md#ec-getting-started-configurations) documents.
```

A list of deprecated and valid ICs/DTs can be found on the [Available regions, deployment templates and instance configurations](cloud://reference/cloud-hosted/ec-regions-templates-instances.md) page, as well as through the API, using `hide_deprecated` to return valid ICs/DTs. 

For example, to return valid ICs/DTs the following request can be used: `https://api.elastic-cloud.com/api/v1/deployments/templates?region=us-west-2&hide_deprecated=true`. To list only the deprecated ones, this can be used: `https://api.elastic-cloud.com/api/v1/deployments/templates?region=us-west-2&metadata=legacy:true`.

If a deprecated IC/DT is already in use, it can continue to be used. However, creating or migrating to a deprecated IC/DT is no longer possible and will result in a plan failing. In order to migrate to a valid IC/DT, navigate to the **Edit hardware profile** option in the Cloud UI or use the [Deployment API](https://www.elastic.co/docs/api/doc/cloud/operation/operation-migrate-deployment-template).

In addtion, you can refer to below information about how these terminologies are referenced. 
* _Deprecated_ is also referenced as _legacy_. 

> * When using APIs, `hide_deprecated=true` is equivalent as `metadata=legacy:true`.

Author: ```suggestion
* When using APIs, `hide_deprecated=true` is equivalent as `metadata=legacy:false`. Or in contrast, using `hide_deprecated=false` is equivalent as `metadata=legacy:true`.
```

Oh I am sorry for that, I must have written a typo here. 🤦 Thank you @claudia-correia  for the very valuable check! 🙏

> * When using APIs, `hide_deprecated=true` is equivalent as `metadata=legacy:true`.

Author: (I made a suggestion based on your comment and committed it)

> * When using APIs, `hide_deprecated=true` is equivalent as `metadata=legacy:true`.

Reviewer 2: Thanks for addressing my comment, Kuni 🙇🏻‍♀️ However, on a second thought, I think we should be a bit more explicit about this, as it may cause confusion to customers. 

My statement above ("Using `hide_deprecated=true` would be equivalent to using `metadata=legacy:false`") is not entirely true, actually. Sorry about that. Let me try to make everything clearer:
- We consider a DT to be deprecated/legacy whenever it contains a `legacy: true` metadata entry.
- If the DT does **not** contain a `legacy` metadata entry **OR** it contains that entry set to `false`, then the DT is considered active/valid (i.e., not deprecated/legacy).
- When using the `metadata=legacy:true` query parameter, this will return all legacy/deprecated DTs.
- When using the `hide_deprecated=true` query parameter, this will return all active/valid DTs.
- Using `metadata=legacy:false` won't actually return anything because usually we don't set the `legacy` field to `false`, we simply omit that field from the metadata. This is why using `hide_deprecated=true` is not actually equivalent to using `metadata=legacy:false`. Sorry for the confusion here.

With this being said, I think we should only mention the three possible scenarios in the docs:
- Using `metadata=legacy:true` to get **legacy/deprecated** DTs.
- Using `hide_deprecated=true` to get **active/valid** DTs.
- Not using any of the query parameters above to get **all** DTs.
  - In this case, we can check the presence of `legacy: true` in the metadata entries within the API response, to verify if an IC/DT is deprecated or not.

I also wonder if it would be worth it to add to the docs that "The above API requests return a list of DTs and the respective ICs referenced within each DT".

* Using `metadata=legacy:false` is not available. Therefore, to verify an IC being deprecated or not, check the presence of `legacy` field in the above API response. 

> * Field `hidden` from API response is not used to present whether an IC being deprecated / legacy or not. 

Author: ```suggestion
* Field `hidden` from API response is completely unrelated to whether an IC is deprecated / legacy or not.
```

Thank you @claudia-correia it makes much better and logical sense!

### API access

Use the [Get deployment templates API](https://www.elastic.co/docs/api/doc/cloud/operation/operation-get-deployment-templates-v2) with query parameters like `hide_deprecated` to retrieve valid ICs and DTs. This API request returns a list of DTs along with the respective ICs referenced within each DT.

For example, 
* To return valid ICs/DTs the following request can be used: `https://api.elastic-cloud.com/api/v1/deployments/templates?region=us-west-2&hide_deprecated=true`. 
* To list only the deprecated ones, this can be used: `https://api.elastic-cloud.com/api/v1/deployments/templates?region=us-west-2&metadata=legacy:true`.

If a deprecated IC/DT is already in use, it can continue to be used. However, creating or migrating to a deprecated IC/DT is no longer possible and will result in a plan failing. In order to migrate to a valid IC/DT, navigate to the **Edit hardware profile** option in the Cloud UI or use the [Deployment API](https://www.elastic.co/docs/api/doc/cloud/operation/operation-migrate-deployment-template).

::::{note}

> Deployment with Elastic stack version prior to 7.10 does not support hardware profile change {{ecloud}} console and API. If you want to make change on hardware profile, upgrading to version 7.10 and onwards is required.

Reviewer: ```suggestion
Deployments using {{stack}} versions prior to 7.10 do not support changing the hardware profile through the {{ecloud}} console or API. To change the hardware profile, first upgrade to version 7.10 or later.
```

::::

In addtion, you can refer to below information about how these terminologies are referenced. 
* _Deprecated_ is also referenced as _legacy_. 

> * Using `metadata=legacy:true` to get legacy/deprecated DTs.
> * Using `hide_deprecated=true` to get active/valid DTs.
> * Not using any of the query parameters above to get all DTs. In this case, check the presence of `legacy: true` in the metadata entries within the API response, to verify if an IC/DT is deprecated or not.

Reviewer 2: ```suggestion
* Using the `metadata=legacy:true` query parameter will return only legacy/deprecated DTs.
* Using the `hide_deprecated=true` query parameter will return only valid DTs.
* Not using any of the query parameters above will return all DTs. In this case, check the presence of `legacy: true` in the `metadata` entries within the API response, to verify if an IC/DT is deprecated or not.
```
